### DISCUSSION TITLE
Is Shadow a good fit for our evaluation requirements? Does a mixnet plugin make sense?

### DISCUSSION THREAD

**Discussion by numbleroot** [shadow/shadow#764] [Q&A] [answered]
Hi everyone,

Thanks for developing and maintaining such a great tool. We (/cc @gunesacar) are currently working on a Tor study and are considering to use Shadow for our experiments. However, after reading the available documentation, we have some remaining questions about whether Shadow can be used for our experimental setup. We would appreciate your input on this.

Our setup involves all nodes involved in a Hidden Service access scenario: a client, a guard, a middle and an exit relay, HSDirs to respond to descriptor lookups, and finally the Hidden Service itself (a web
application or a simple web page).

1) High-level: Is this an experiment setup well-suited to be recreated within Shadow?

2) Does Shadow allow us to measure the memory and CPU overhead of an attack on the guard nodes?

3) Is Shadow able to emulate the Hidden Service descriptor lookup procedure (involving the HSDirs)?

4) One of the metrics to evaluate will be time-to-completion of an attack. Given that Shadow simulates time, is this a deal breaker? How would a simulated time duration be translated into "real-world durations"?

5) Can you think of any Shadow limitations that might hinder conducting our experiments with it?

We have an additional, unrelated question: Do you think it would be possible to build a mixnet plugin (Chaumian, re-encryption, etc.) for Shadow? That is, a plugin that simulates a mixnet that works very differently from Tor (no circuits, intentional delaying of messages, different network topologies, etc.). Do you see any obstacles in the way Shadow works that would prevent a mixnet plugin from making sense?

We would appreciate your input on these questions. Thanks in advance!

**Comment by robgjansen** [comment, marked as answer]
Hello!

> 1. High-level: Is this an experiment setup well-suited to be recreated within Shadow?

I don't think there is anything about what you described so far that would indicate that such an experiment would not work. Generally, any functionality that is provided in the tor network daemon (little-t tor, not Tor browser) should be available when running tor Shadow. Specifically, hidden services are a less-often used feature when running Shadow experiments, but AFAIK they do work correctly. We have an example configuration of a ~10 node Shadow experimentation configuration that includes a hidden service; you could play around with this small-scale example to make sure it has what you need before fully committing to Shadow:

https://github.com/robgjansen/shadow-plugin-tor/blob/example-fixup/resource/shadowtor-minimal-config.tar.xz

> 2. Does Shadow allow us to measure the memory and CPU overhead of an attack on the guard nodes?

The current version of Shadow (Shadow v1.14.x) runs in a single process, with all nodes (e.g., Tor nodes) running as plugins that are dynamically loaded at runtime. From the OS perspective, you can measure the CPU and memory overheads from running a full experiment, with and without the attack. These measurements will be fairly precise.

Measuring the memory and CPU of individual nodes inside of an experiment will admittedly be a bit more challenging but not impossible. We did this for the Sniper Attack paper. This is one area that we know needs some improvement.

CPU:

It amounts to tracking the time the CPU spends inside of the plugin code (starting a timer when we enter the plugin code, and stoping the timer when we exit the plugin code). This is done for each process running on each host. When the experiment ends, the cumulative values of these timers are logged to file. You could use this result as a full measure of CPU time for a node. You could also alter the code to log the state of the timers more often, e.g., every second.

Going one step further is making sure that if a process is consuming a lot of CPU time dealing with your attack, then that should "slow down" its processing of other tasks like network communication. We currently have an experimental mode that does this (you turn it on with a flag to the shadow binary), but it extends the amount of time it takes to complete an experiment due to increased churn on the master event queues. (Essentially, whenever the node CPU uses up over a threshold time of processing, say 1 second, then we have to move back all other events in its queue by 1 second. This can get costly.)

RAM:

There is another flag that will turn on a memory tracker. This causes Shadow to interpose on malloc and free and using those to count the current amount of RAM in use on a node.

I haven't used these features in a bit, so I'm not sure if they will work 100% out of the box or not.

> 3. Is Shadow able to emulate the Hidden Service descriptor lookup procedure (involving the HSDirs)?

Yes, because we are directly executing Tor code, the same lookup procedure would be executed whether running in Shadow or not.

> 4. One of the metrics to evaluate will be time-to-completion of an attack. Given that Shadow simulates time, is this a deal breaker? How would a simulated time duration be translated into "real-world durations"?

Simulated time is a direct representation of real time. If it takes 10 seconds of simulated time in Shadow to download a file, then we expect it would take 10 seconds of real time to download that file outside of Shadow. Therefore, I expect that the time to complete an attack would be accurately represented in Shadow. The thing that would cause inaccuracy is the CPU model as described above; if the CPU model is turned off, then all CPU processing is essentially done "for free" and that may or may not be a problem depending on the types of attacks you want to explore.

> 5. Can you think of any Shadow limitations that might hinder conducting our experiments with it?

Language: Shadow currently works best with plugins that are written in C/C++. Python/Java/etc. will likely not work well with Shadow v1.x.x. We are working on a new architecture and Shadow v2 will support any language, but currently this is a limitation.

Internet: Shadow currently uses a network graph to model Internet endpoints (nodes) and paths (edges). We use this to model latencies on the edges between various vantage points on the Internet. Simulation nodes get placed on network endpoints, and then the paths tell us how long packets take to arrive at another endpoint. This is fairly high level, there are no routing protocols autonomous systems, etc. This is usually OK for application-level experiments, but you may have problems if your experiment depends on routing-level semantics.

Forking: Shadow can handle applications that uses multiple threads, and it can run multiple processes per host. But if your software forks, that will cause problems and we currently don't have support for it.

> We have an additional, unrelated question: Do you think it would be possible to build a mixnet plugin (Chaumian, re-encryption, etc.) for Shadow? That is, a plugin that simulates a mixnet that works very differently from Tor (no circuits, intentional delaying of messages, different network topologies, etc.). Do you see any obstacles in the way Shadow works that would prevent a mixnet plugin from making sense?

Shadow is a replacement for the networking layer, i.e., the Internet. It is designed to directly execute plugins like Tor. You could create a new plugin that has its own application logic, and it should run fine in Shadow, as long as it is written to run in a Linux environment. In fact, we have done this for some other applications like the traffic generator we use in our Tor experiments:
https://github.com/shadow/tgen

A main limitation would be choice of programming language, as described above.

Hope that helps! ~Rob

**Comment by numbleroot** [comment]
Hi Rob,

Thanks a lot for the very detailed response. Great to hear that Shadow appears to fit our evaluation requirements. We will definitely check out the example configuration you linked.

Out of curiosity: Is there a feature roadmap and a planned release time for the Shadow v2 release?

Thanks again!

**Comment by robgjansen** [comment]
> Out of curiosity: Is there a feature roadmap and a planned release time for the Shadow v2 release?

We are still in the research/prototyping stage, so we don't have a full roadmap yet for a stable release of Shadow v2. However, because we have a lot of changes planned for v2, it is very likely that we will be sharing a v2 development branch as soon as we think it is stable enough to do so. Although uncertain, I suspect that will happen sometime in the next few months. The dev branch will help those that are OK trying out code that is less well tested in advance of a formal v2 stable release.

**Comment by robgjansen** [comment]
Also, we will likely post the feature roadmap on the milestones page, once we start mapping out what exactly we want a v2 release to look like:
https://github.com/shadow/shadow/milestones

**Comment by numbleroot** [comment]
Sounds good, thanks for sharing this! Will certainly check out the roadmap once it exists.
